### PR TITLE
FINERACT-1095 Added status parameter in Loans API

### PR DESCRIPTION
**Pull request by thesmallstar**
This is in continuation to #1207
1. support for new parameter "status" to loans is added.
2. It can have multiple values example: staus=active&status=approved
3. Validations are also added.

### CODE REVIEW

**Comment by thesmallstar** [comment]
@vorburger  @ptuomola  @luckyman20 
As soon as this is merged, we can completely remove sqlsearch and the PRs #1171 and #1123 can be merged with minor changes.
And we will be done with removing all SQL string concatenations from our code :D

**Comment by ptuomola** [comment]
@thesmallstar the code looks great to me - thanks!

Can you add some intergration tests as well so we have this feature covered and know it works? e.g. search with one status, search with multiple statuses, and search with invalid status value?

**Comment by luckyman20** [comment]
@thesmallstar It is not possible in community-app that we can send the same parameter more than once with the request, please look into this.

**Comment by vorburger** [comment]
@thesmallstar are you on this? :smiling_imp:

**Comment by thesmallstar** [comment]
@vorburger  yes yes, I am in conversion with @luckyman20 we are trying to resolve the frontend part. We should confirm it by today, And then find the best way for this.

**Comment by thesmallstar** [comment]
@vorburger  @ptuomola  taking exceptionally long time to write tests since it is really hard to understand how loans work :(

**Comment by vorburger** [comment]
> @vorburger @ptuomola taking exceptionally long time to write tests since it is really hard to understand how loans work :(

if you have a more specific question, then perhaps @ptuomola (likely better than me!) can answer.

but if it's just writing integration tests for this that's very significant additional work, than perhaps it wouldn't be the end of the world to get this merged without additional new ITs already (but it can break existing ITs, of course), and then perhaps later add more test coverage for it, when (if) you have more time.

**Comment by thesmallstar** [comment]
@vorburger It would be nice if we could merge this rn, Since this is blocking the work for removing SQL String concatenation, I will have enough time and I am making an issue so that I don't forget the IT for this ;) 

Edit: Don't MERGE THIS RN :(

**Comment by thesmallstar** [comment]
@luckyman20 and I, are still not able to figure out how to do this on the frontend, I will try to get it done tomorrow, after which we can merge this.

**Comment by ptuomola** [comment]
> @vorburger @ptuomola taking exceptionally long time to write tests since it is really hard to understand how loans work :(

@thesmallstar Hopefully you can reuse the code from ClientLoanIntegrationTest? That creates and checks lots of loans in different statuses and has utils for doing that - so creating additional loans and putting them to different statuses and then testing searching should just be a matter of some copy & paste from existing tests.

**Comment by thesmallstar** [comment]
@ptuomola  @vorburger , From my discussions with @luckyman20 we came to the conclusion that we cannot send two parameters from the community-app, do you see any other solution to solve this one? 
We need our API to accept a list of strings...

**Comment by ptuomola** [comment]
@thesmallstar did you see my earlier suggestion in the jira ticket on how to do this?

**Comment by vorburger** [comment]
@thesmallstar the suggestion from @ptuomola in the last comment on FINERACT-1095 to use e.g. `"loanStatus": ["all","active","overpaid"], "loanProducts": ["65","97","73","82"], "offices": ["1","10","100"]` syntax makes a lot of sense, to me.. will you do the needful here? It would be cool if we could wrap up FINERACT-1095!

**Comment by thesmallstar** [comment]
@vorburger  yup I will fix this in a day or two. 
@ptuomola  thanks for the suggestion!!

**Comment by vorburger** [comment]
@thesmallstar friendly ping, update? (as well as now have to rebase and fix merge conflict)

**Comment by thesmallstar** [comment]
@vorburger sorry for the delay, I am in between of my college mid-semester exams, I will continue working as soon as I am done with it(hopefully no sooner than 19th Sept).

**Comment by vorburger** [comment]
@thesmallstar hope your exams went well! Whenever you have time to pick this up, we appreciate your contributions, very much!

**Comment by thesmallstar** [comment]
@vorburger  yes already working on it!

**Comment by thesmallstar** [comment]
@vorburger @ptuomola @luckyman20 Thank you for waiting a long time for these here is the update:

1. As mentioned by @ptuomola  for using the implementation ["a","b"] that cannot be really used here because the request type is not JSON, but we can do something similar.
2. The API now takes one string and works for both: examples:
                    a. "active"
                    b. "active,approved"
Example URLS: 
https://localhost:8443/fineract-provider/api/v1/loans?status=active,approved&tenantIdentifier=default&pretty=true

3. We will merge this now, I will write the Integration tests and I have also made an issue so that we don't forget to do it. https://issues.apache.org/jira/browse/FINERACT-1160

**Comment by vorburger** [comment]
@thesmallstar sounds like a plan, and OK for me to add the IT in a follow up. But the build on this is not passing, did you see? BTW watch the related discussion over on https://github.com/openMF/community-app/issues/3275.

**Comment by thesmallstar** [comment]
@vorburger oops, I missed to ping you after I pushed new changes.
I think we can merge this now?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/LoanStatus.java
Comment (vorburger) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/LoanStatus.java:93]: @thesmallstar I'm probably missing something, but isn't this just `LoanStatus.valueOf(statusString)`? If you're doing this only because of `equalsIgnoreCase()` then wouldn't a `LoanStatus.valueOf(statusString.toUpperCase())` suffice and have the same effect? If you want an invalid status String to return LoanStatus.INVALID, you may have to `catch (IllegalArgumentException e)` .. (I think; not 100% sure, do check). Actually that is debatable.. wouldn't you rather that an invalid status filter passed through the REST API from a client lead to an error, than be silently ignored?

**Review by vorburger — CHANGES_REQUESTED** [review]

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanReadPlatformServiceImpl.java
Comment (thesmallstar) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanReadPlatformServiceImpl.java:305]: @vorburger  here, we are checking for all strings that were given to us and make sure none of them returns an Invalid status if it does we return a PlatformApiDataValidationException, does this answer your question? (I might have missed something?)

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/LoanStatus.java
Comment (thesmallstar) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/LoanStatus.java:93]: For the first part, yes, they would have the same effect. I copied this implementation from some other enum and hence the code looks like this and not how you suggested. Do you see any advantage of one over the other?

2nd part answer in a comment below.

Comment (ptuomola) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/LoanStatus.java:93]: Just another point: given this validation seems to be a common pattern in the enums, I've added a method in the DataValidatorBuilder (isOneOfEnumValues()) to check if the value provided is valid for a specific enum (and report back with correct PlatformApiDataException if it isn't).

So one way to handle these is in the validation phase check the values using that, after which it should be safe to use valueOf() without getting an exception. 

So as far as I can see there are three ways to do the same thing:

- custom valueOf method with error handling to avoid an exception (as currently in many places in the code), and then check for invalid status afterwards - i.e. the current logic in this PR

- call valueOf and catch the exception if the value is invalid, and rethrow as PlatformApiDataException

- call DataValidatorBuilder.isOneOfEnumValues() to validate and then valueOf to convert

Not sure there's a significant advantage of one way over another...

**Comment by vorburger** [comment]
@thesmallstar @ptuomola where were we with this one? Good to merge as-is, or updates needed? (Perhaps best to rebase it now.)

**Comment by ptuomola** [comment]
@vorburger based on the above discussions, my suggestion would still be to change the validation to isOneOfEnumValues() and the conversion to be done with LoanStatus.valueOf(statusString.toUpperCase()) as you suggest. That would avoid a lot of "boilerplate" code.

**Comment by thesmallstar** [comment]
@ptuomola @vorburger thanks, on it :) 
Sorry for delaying this long, I have been juggling between college exams, assignments and health :( 
I will be done with my finals on 28th and have a month-long holiday, and I would have plenty of time to commit daily!!
PS: I miss the summer days and lot of "productive" discussions with @vorburger @ptuomola and @xurror  :D

**Comment by vorburger** [comment]
> Sorry for delaying this long, I have been juggling between college exams, assignments and health :(

Don't say Sorry! You have no "obligation" to continue contributing to this project (after your GSoC) - no stress! Best wishes for health.

> I will be done with my finals on 28th and have a month-long holiday, and I would have plenty of time to commit daily!!

But we are of course excited to hear from you! If you can make the time, we always love to hear from you.

**Pull request closed without merging**
